com.sap.cloud.lm.sl.cf.process/src/main/java/com/sap/cloud/lm/sl/cf/process/steps/DetermineDesiredStateAchievingActionsStep.java

```diff
@@ -46,24 +55,39 @@ protected StepPhase executeStep(ExecutionWrapper context) {
     }
 
     private StepPhase attemptToExecuteStep(ExecutionWrapper execution) {
-        CloudApplication app = StepsUtil.getApp(execution.getContext());
-        ApplicationStartupState currentState = computeCurrentState(execution, app);
-        getStepLogger().debug(Messages.CURRENT_STATE, app.getName(), currentState);
+        String appName = StepsUtil.getApp(execution.getContext())
+            .getName();
+        CloudControllerClient client = execution.getControllerClient();
+        CloudApplication app = client.getApplication(appName);
+        ApplicationStartupState currentState = computeCurrentState(app);
+        getStepLogger().debug(Messages.CURRENT_STATE, appName, currentState);
         ApplicationStartupState desiredState = computeDesiredState(execution.getContext(), app);
-        getStepLogger().debug(Messages.DESIRED_STATE, app.getName(), desiredState);
+        getStepLogger().debug(Messages.DESIRED_STATE, appName, desiredState);
 
         Set<ApplicationStateAction> actionsToExecute = getActionsCalculator(execution.getContext()).determineActionsToExecute(currentState,
             desiredState);
-        getStepLogger().debug(Messages.ACTIONS_TO_EXECUTE, app.getName(), actionsToExecute);
+        addStageStateIfNeeded(client, app, actionsToExecute);
+        getStepLogger().debug(Messages.ACTIONS_TO_EXECUTE, appName, actionsToExecute);
 
         StepsUtil.setAppStateActionsToExecute(execution.getContext(), actionsToExecute);
         return StepPhase.DONE;
     }
 
-    private ApplicationStartupState computeCurrentState(ExecutionWrapper execution, CloudApplication app) {
-        CloudControllerClient client = execution.getControllerClient();
+    private void addStageStateIfNeeded(CloudControllerClient client, CloudApplication cloudApplication,
+        Set<ApplicationStateAction> actionsToExecute) {
+        PlatformType platformType = configuration.getPlatformType();
+        if (platformType != PlatformType.CF) {
+            return;
+        }
+        ApplicationStager applicationStager = ApplicationStagerFactory.createApplicationStager(platformType);
+        if (!applicationStager.isApplicationStagedCorrectly(client, cloudApplication)) {
+            actionsToExecute.add(ApplicationStateAction.STAGE);
```

Comment:
I'm beginning to think that it would be a good idea if we were to print a warning to the user saying that his application was not staged correctly last time and that we're going to restage it. It would also be good to log the most recent build, so that if there's an open ticket we can investigate through the logs downloaded by "cf dmol".